ipn/ipnlocal/profiles.go

```diff
@@ -0,0 +1,458 @@
+	return pm.prefs
+}
+
+// NewProfileManager creates a new ProfileManager using the provided StateStore.
+// It also loads the list of known profiles from the StateStore.
+// If a state key is provided, it will be used to load the current profile.
+func NewProfileManager(store ipn.StateStore, logf logger.Logf, stateKey ipn.StateKey) (*ProfileManager, error) {
+	return newProfileManagerWithGOOS(store, logf, stateKey, runtime.GOOS)
+}
+
+func readAutoStartKey(store ipn.StateStore, goos string) (ipn.StateKey, error) {
+	startKey := ipn.CurrentProfileStateKey
+	if goos == "windows" {
+		// When tailscaled runs on Windows it is not typically run unattended.
+		// So we can't use the profile mechanism to load the profile at startup.
+		startKey = ipn.ServerModeStartKey
+	}
+	autoStartKey, err := store.ReadState(startKey)
+	if err != nil && err != ipn.ErrStateNotExist {
+		return "", fmt.Errorf("calling ReadState on state store: %w", err)
+	}
+	return ipn.StateKey(autoStartKey), nil
+}
+
+func readKnownProfiles(store ipn.StateStore) (map[ipn.ProfileID]*ipn.LoginProfile, error) {
+	var knownProfiles map[ipn.ProfileID]*ipn.LoginProfile
+	prfB, err := store.ReadState(ipn.KnownProfilesStateKey)
+	switch err {
+	case nil:
+		if err := json.Unmarshal(prfB, &knownProfiles); err != nil {
+			return nil, fmt.Errorf("unmarshaling known profiles: %w", err)
+		}
+	case ipn.ErrStateNotExist:
+		knownProfiles = make(map[ipn.ProfileID]*ipn.LoginProfile)
+	default:
+		return nil, fmt.Errorf("calling ReadState on state store: %w", err)
+	}
+	return knownProfiles, nil
+}
+
+func newProfileManagerWithGOOS(store ipn.StateStore, logf logger.Logf, stateKey ipn.StateKey, goos string) (*ProfileManager, error) {
+	if stateKey == "" {
+		var err error
+		stateKey, err = readAutoStartKey(store, goos)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	knownProfiles, err := readKnownProfiles(store)
+	if err != nil {
+		return nil, err
+	}
+
+	pm := &ProfileManager{
+		store:         store,
+		knownProfiles: knownProfiles,
+		logf:          logf,
+	}
+
+	if stateKey != "" {
+		for _, v := range knownProfiles {
+			if v.Key == stateKey {
+				pm.currentProfile = v
+			}
+		}
+		if pm.currentProfile == nil {
+			if suf, ok := strs.CutPrefix(string(stateKey), "user-"); ok {
+				pm.currentUserID = suf
+			}
+			pm.isNewProfile = true
+		} else {
+			pm.currentUserID = pm.currentProfile.LocalUserID
+		}
+		prefs, err := pm.loadSavedPrefs(stateKey)
+		if err != nil {
+			return nil, err
+		}
+		if err := pm.setPrefsLocked(prefs); err != nil {
+			return nil, err
+		}
+	} else if len(knownProfiles) == 0 && goos != "windows" {
+		// No known profiles, try a migration.
+		if err := pm.migrateFromLegacyPrefs(); err != nil {
+			return nil, err
+		}
+	} else {
+		pm.prefs = emptyPrefs
+	}
+
+	return pm, nil
+}
+
+func (pm *ProfileManager) migrateFromLegacyPrefs() error {
+	pm.NewProfile()
+	k := ipn.LegacyGlobalDaemonStateKey
+	switch runtime.GOOS {
+	case "ios", "darwin":
```

Comment:
`tailscaled` on macOS does not use `ipn-go-bridge`, it should still read from `LegacyGlobalDaemonStateKey`

Reply:
Good catch! I thought about it last night but promptly forgot :)